pkg/tfbridge/schema.go

```diff
@@ -461,13 +461,24 @@ const metaKey = "__meta"
 // MakeTerraformResult expands a Terraform state into an expanded Pulumi resource property map.  This respects
 // the property maps so that results end up with their correct Pulumi names when shipping back to the engine.
 func MakeTerraformResult(state *terraform.InstanceState,
-	tfs map[string]*schema.Schema, ps map[string]*SchemaInfo) resource.PropertyMap {
+	tfs map[string]*schema.Schema, ps map[string]*SchemaInfo) (resource.PropertyMap, error) {
 	var outs map[string]interface{}
 	if state != nil {
 		outs = make(map[string]interface{})
 		attrs := state.Attributes
+
+		reader := &schema.MapFieldReader{
+			Schema: tfs,
+			Map:    schema.BasicMapReader(attrs),
+		}
 		for _, key := range flatmap.Map(attrs).Keys() {
-			outs[key] = flatmap.Expand(attrs, key)
+			res, err := reader.ReadField([]string{key})
+			if err != nil {
+				return nil, err
+			}
+			if res.Value != nil {
```

Comment:
This will only happen for keys that are not present in the resource's schema. One notable field in this category is the `id` field, which we will no longer populate. I do not believe that this should be problematic--the only values we expose through the SDKs today are values that are present in the schema, and we communicate the value of the `id` field out of band--but I am open to the idea of backfilling properties that are present in the flatmap but not in the schema. Opinions?

Reply:
(FWIW, it would be _very_ easy to do something as simple as retain the `id` field)

Reply:
For data-sources, I recall we had issues with not populating `id` in outputs?  Will this impact that?

Am I correct that you believe these changes result in no user-visible change in behaviour?  If there is expected user-visibile change in behaviour - what is it exactly?

Reply:
> For data-sources, I recall we had issues with not populating id in outputs? Will this impact that?

This will not impact that, no: we explicitly set the ID property in our implementation of `Invoke`.

> Am I correct that you believe these changes result in no user-visible change in behaviour? If there is expected user-visibile change in behaviour - what is it exactly?

I believe that it is exceedingly unlikely that these changes will cause any changes to user-visible behavior besides the following:
- The `id` property will no longer be present in resource states. This will show up during a `refresh` as a removal. If this is undesirable (and it may well be), it is easy enough to restore this specific property.
- Maps with keys that contain characters that are semantically meaningful to flatmap will be properly marshaled/unmarshaled (this is the bugfix)

It is _possible_ that other properties that are not present in the schema will be missing, but those are only exposed to users via the checkpoint itself (not via the SDKs).

Reply:
>  If this is undesirable (and it may well be), it is easy enough to restore this specific property.

Yes - that sounds undesirable.  Let's restore that the `id` property.

Reply:
Done.